### PR TITLE
add csr sample op

### PR DESCRIPTION
**Pull request by aksnzhy**
This pr add the graph sampling op, which will be used in the DGL project.
@zheng-da Please review this pr. Thanks!

Benchmark code:

```
import mxnet as mx
import numpy as np
import random
import time

vertex_num = 5000000
num_neighbor = 16
max_num_vertices = 1000000

# Create a big csr
def create_big_csr_graph():
    a = mx.nd.load('../5_5_csr.nd')[0]
    return a

# Create a list of random seed [0, max_id]
def create_seed(num_seed, max_id):
	res = []
	for x in range(num_seed):
		res.append(random.randint(0, max_id))
	return res

def benchmark_non_uniform(graph, seed_size, hops, neighs, max_vertices):
	for seed_num in seed_size:
		seed_list = []
		for n in range(32):
			seed = create_seed(seed_num, vertex_num-1)
			seed_ndarray = mx.nd.array(seed, dtype='int64')
			seed_list.append(seed_ndarray)
		prob = mx.nd.random.uniform(0, 1, shape=(vertex_num,))
		start = time.time()
		out = mx.nd.contrib.dgl_csr_neighbor_non_uniform_sample(graph, prob, *seed_list, 
			num_hops=hops, 
			num_neighbor=neighs, 
			max_num_vertices=max_vertices)
		out[0].wait_to_read()
		elapsed = (time.time() - start)
		ver_count = 0
		edge_count = 0
		for i in range(31):
			ver_count += out[i].size
			edge_count += out[i+32].data.size
		print("time: " + str(elapsed) + " (s)" +
			  " thread: 32" + 
			  " seed size: " + str(seed_num) + 
			  " num_hops: " + str(hops) + 
			  " num neighbors: " + str(neighs) +
			  " max vertices: " + str(max_vertices) +
			  " sample_num_ver: " + str(ver_count) +
			  " sample_num_edge: " + str(edge_count))

def benchmark_uniform(graph, seed_size, hops, neighs, max_vertices):
	for seed_num in seed_size:
		seed_list = []
		for n in range(32):
			seed = create_seed(seed_num, vertex_num-1)
			seed_ndarray = mx.nd.array(seed, dtype='int64')
			seed_list.append(seed_ndarray)
		start = time.time()
		out = mx.nd.contrib.dgl_csr_neighbor_uniform_sample(graph, *seed_list, 
			num_hops=hops, 
			num_neighbor=neighs, 
			max_num_vertices=max_vertices)
		out[0].wait_to_read()
		elapsed = (time.time() - start)
		ver_count = 0
		edge_count = 0
		for i in range(32):
			ver_count += out[i].size
			edge_count += out[i+32].data.size
		print("time: " + str(elapsed) + " (s)" +
			  " thread: 32" + 
			  " seed size: " + str(seed_num) + 
			  " num_hops: " + str(hops) + 
			  " num neighbors: " + str(neighs) +
			  " max vertices: " + str(max_vertices) +
			  " sample_num_ver: " + str(ver_count) +
			  " sample_num_edge: " + str(edge_count))

if __name__ == "__main__":
	graph = create_big_csr_graph()
	seed_size = [1000, 2000, 3000, 4000, 8000]
	print("----------------------------------Uniform Benchmark----------------------------------")
	benchmark_uniform(graph, seed_size, 1, num_neighbor, max_num_vertices)
	benchmark_uniform(graph, seed_size, 2, num_neighbor, max_num_vertices)
	print("--------------------------------Non-Uniform Benchmark--------------------------------")
	benchmark_non_uniform(graph, seed_size, 1, num_neighbor, max_num_vertices)
	benchmark_non_uniform(graph, seed_size, 2, num_neighbor, max_num_vertices)
```

### CODE REVIEW

File: src/operator/contrib/dgl_graph.cc [hunk 1]
```diff
@@ -35,6 +35,833 @@
 namespace mxnet {
 namespace op {
 
+typedef int64_t dgl_id_t;
+
+////////////////////////////// Graph Sampling ///////////////////////////////
+
+unsigned int seed = 123;
```
Comment (zheng-da) [inline src/operator/contrib/dgl_graph.cc]: we should't fix a seed like this.

File: src/operator/contrib/dgl_graph.cc [hunk 2]
```diff
@@ -35,6 +35,833 @@
...
+                       num_neighbor,
+                       &tmp_sampled_src_list,
+                       &tmp_sampled_edge_list);
+      }
+      neigh_mp.insert(std::pair<dgl_id_t, neigh_list>(dst_id,
+        neigh_list(tmp_sampled_src_list, tmp_sampled_edge_list)));
+      num_edges += tmp_sampled_src_list.size();
+      bool exit = false;
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        auto ret = sub_ver_mp.insert(tmp_sampled_src_list[i]);
```
Comment (zheng-da) [inline src/operator/contrib/dgl_graph.cc]: merge `sub_ver_mp` and `layer`?

File: src/operator/contrib/dgl_graph.cc [hunk 3]
```diff
@@ -35,6 +35,833 @@
...
+  }
+  // The last element stores the actual
+  // number of vertices in the subgraph.
+  out[max_num_vertices] = sub_ver_mp.size();
+  // Copy sub_probability
+  if (sub_prob != nullptr) {
+    for (dgl_id_t i = 0; i < max_num_vertices; ++i) {
+      dgl_id_t idx = out[i];
+      if (idx != -1) {
+        sub_prob[i] = probability[idx];
```
Comment (zheng-da) [inline src/operator/contrib/dgl_graph.cc]: we need to have the probability that a neighbor is sampled from the the neighbor list. In other words, it should be normalized for the neighbor list.

File: tests/python/unittest/test_dgl_graph.py [hunk 4]
```diff
@@ -26,6 +26,41 @@
...
+    indices_np = np.array([1,2,3,4,0,2,3,4,0,1,3,4,0,1,2,4,0,1,2,3], dtype=np.int64)
+    indptr_np = np.array([0,4,8,12,16,20], dtype=np.int64)
+    a = mx.nd.sparse.csr_matrix((data_np, indices_np, indptr_np), shape=shape)
+    seed = mx.nd.array([0,1,2,3,4], dtype=np.int64)
+    out = mx.nd.contrib.csr_neighbor_uniform_sample(a, seed, num_args=2, num_hops=1, num_neighbor=2, max_num_vertices=5)
+    sample_id = out[0]
+    count = 0
+    for data in sample_id:
+        if data != -1:
+            count = count + 1
```
Comment (zheng-da) [inline tests/python/unittest/test_dgl_graph.py]: can you double check the sampling for 2-hop neighborhood?

File: tests/python/unittest/test_dgl_graph.py [hunk 5]
```diff
@@ -26,6 +26,41 @@
...
 import unittest
 
+def test_uniform_sample():
+    shape = (5, 5)
+    data_np = np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20], dtype=np.int64)
+    indices_np = np.array([1,2,3,4,0,2,3,4,0,1,3,4,0,1,2,4,0,1,2,3], dtype=np.int64)
+    indptr_np = np.array([0,4,8,12,16,20], dtype=np.int64)
+    a = mx.nd.sparse.csr_matrix((data_np, indices_np, indptr_np), shape=shape)
+    seed = mx.nd.array([0,1,2,3,4], dtype=np.int64)
+    out = mx.nd.contrib.csr_neighbor_uniform_sample(a, seed, num_args=2, num_hops=1, num_neighbor=2, max_num_vertices=5)
```
Comment (zheng-da) [inline tests/python/unittest/test_dgl_graph.py]: we should check every output.

File: tests/python/unittest/test_dgl_graph.py [hunk 6]
```diff
@@ -26,6 +26,41 @@
...
+    seed = mx.nd.array([0,1,2,3,4], dtype=np.int64)
+    out = mx.nd.contrib.csr_neighbor_non_uniform_sample(a, prob, seed, num_args=3, num_hops=1, num_neighbor=2, max_num_vertices=5)
+
+    sample_id = out[0]
+    count = 0
+    for data in sample_id:
+        if data != -1:
+            count = count + 1
+
+    assert (mx.nd.array([count-1], dtype=np.int64)== sample_id[-1])
```
Comment (zheng-da) [inline tests/python/unittest/test_dgl_graph.py]: For the non-uniform case, we should verify the sampled vertices are sampled with the specified probability. But we probably can't write the test in the unit test.

File: src/operator/contrib/dgl_graph.cc (hunk 1 above)
Comment (aksnzhy) [reply src/operator/contrib/dgl_graph.cc]: I think we need to use mxnet-random here.

Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: Are all threads sharing the same random seed?

File: src/operator/contrib/dgl_graph.cc [hunk 7]
```diff
@@ -35,6 +35,833 @@
...
+  shape_1[0] = num_edges;
+  shape_2[0] = max_num_vertices+1;
+  sub_csr.CheckAndAllocData(shape_1);
+  sub_csr.CheckAndAllocAuxData(csr::kIdx, shape_1);
+  sub_csr.CheckAndAllocAuxData(csr::kIndPtr, shape_2);
+  dgl_id_t* val_list_out = sub_csr.data().dptr<dgl_id_t>();
+  dgl_id_t* col_list_out = sub_csr.aux_data(1).dptr<dgl_id_t>();
+  dgl_id_t* indptr_out = sub_csr.aux_data(0).dptr<dgl_id_t>();
+
+  size_t collected_nedges = 0;
```
Comment (zheng-da) [inline src/operator/contrib/dgl_graph.cc]: you need to set indptr_out[0] to be 0.

File: src/operator/contrib/dgl_graph.cc (hunk 2 above)
Comment (aksnzhy) [reply src/operator/contrib/dgl_graph.cc]: Fixed it

File: tests/python/unittest/test_dgl_graph.py (hunk 4 above)
Comment (aksnzhy) [reply tests/python/unittest/test_dgl_graph.py]: Add new test

File: tests/python/unittest/test_dgl_graph.py (hunk 5 above)
Comment (aksnzhy) [reply tests/python/unittest/test_dgl_graph.py]: Fixed it

File: src/operator/contrib/dgl_graph.cc [hunk 8]
```diff
@@ -35,6 +35,833 @@
...
+.describe(R"code(This operator samples sub-graph from a csr graph via an
+uniform probability. 
+For example:
+shape = (5, 5)
+prob = mx.nd.array([0.9, 0.8, 0.2, 0.4, 0.1], dtype=np.float32)
+data_np = np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20], dtype=np.int64)
+indices_np = np.array([1,2,3,4,0,2,3,4,0,1,3,4,0,1,2,4,0,1,2,3], dtype=np.int64)
+indptr_np = np.array([0,4,8,12,16,20], dtype=np.int64)
+a = mx.nd.sparse.csr_matrix((data_np, indices_np, indptr_np), shape=shape)
+seed = mx.nd.array([0,1,2,3,4], dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/operator/contrib/dgl_graph.cc]: the seed is necessary for this sampling algorithm.

[the remark this answers is not in the document]
Comment (aksnzhy) [reply src/operator/contrib/dgl_graph.cc]: The seed is not used for random operation. Actually, it is seed-vertex, which is the initial vertex used for sampling.

**Pull request merged**